feat: first first draft (#4)
kawamurakazushi/presentation-react-hooks: slides.md

class: middle, center

## The Context API

---

class: middle, center

## The Hooks API

---

class: middle

> Hooks are a new feature proposal that lets you use state and other React features without writing a class. They’re currently in **React v16.7.0-alpha**

---

### What does Hook solve?

> React Component is great to organize large UI into small, independent, and reusuable piece.

Reviewer: pieces?

> However, it is quite challenging to break down complex components, since the logic is **stateful** and can't be extracted to a function or another component.

Reviewer: or _to_ an other component

---

### Complex components, with logic all over the lifecyle

```typescript
class App extends React.Component {
  constructor(props) {
    super(props);
    this.state = { name: "Kawamura", like: "curry" };
  }
  componentDidMount() {
    fetchA();
    fetchB();
    subscribe();
  }
  componentWillUpdate(props) {
    if (props.a !== this.props.a) fetchA();
    if (props.b !== this.props.b) fetchB();
  }
  componentWillUnmount() { unSubscribe(); }
  handleNameChange(event) { this.setState({ name: event.target.value }); }
  render() { ... }
}
```

> Each _lifecycle methods_ is consisted with **unrelated logics**, and the related code that changes together gets **split apart**.

Reviewer: ```suggestion
The business logic, in all _lifecycle methods_ are _internally_ **inconsistent**, `componentDidMount` for instance will fetch `A` and `B` here.

Also, if we wanted to extract the methods, the logic would be **split apart**.

Basically, instead of splitting the code for technical reasons, we want to split by concern.
```

Very difficult to maintain 😇

---

### Some Solutions.

- Context
- Render props
- Higher Order Components

> These patterns requires you to restructure your components.  

Reviewer: ```suggestion
These patterns require you to restructure your components.  
```

> Which can be **tedious**, make code harder to follow, and also creates **false hierarchy**.

Reviewer: ```suggestion
Which can be **tedious**, make the code harder to follow, and also creates **false hierarchy**.
```

```typescript
<ProviderA>
  <ProviderB>
    <SomeRenderProps>{({ data }) => <div>{data}</div>}</SomeRenderProps>
  </ProviderB>
  <RenderProps>{({ data }) => <div>{data}</div>}</SomeRenderProps>
</ProviderA>
```

Wouldn't it be easier to have **one** common way to share logic between components?  
... functions?

Functions _can't have local state_.

---

### Reusing stateful logic between Components.

> **Hooks** are greak solution for solving these problems.

Reviewer: ```suggestion
**Hooks** are a great solution for solving these problems.
```

> Hooks let us use React features from a function.  

Reviewer: ```suggestion
Hooks let us use React features, like statefulness, effects, etc..., from a function.  
```

> It can process throught `state`, `context`, and `lifecycle`.

Reviewer: ```suggestion

```

> We can extract and share the logic inside a component into reusable independent function without changing your component hireachy.

Reviewer: ```suggestion
We can extract and share the logic inside a component into reusable independent functions without changing your component hierarchy.
```

---

### Example - Without Hooks

```typescript
class App extends React.Component {
  constructor(props) {
    super(props);
    this.state = { name: "Kawamura", favoriteFood: "curry" };
    this.handleNameChange = this.handleNameChange.bind(this);
    this.handleFavoriteFoodChange = this.handleFavoriteFoodChange.bind(this);
  }
  handleNameChange(event) {
    this.setState({ name: event.target.value });
  }
  handleFavoriteFoodChange(event) {
    this.setState({ favoriteFood: event.target.value });
  }
  render() {
    const { name, favoriteFood } = this.state;
    return (
      <>
        <input value={name} onChange={handleNameChange} />
        <input value={favoriteFood} onChange={handleFavoriteFood} />
      </>
    );
  }
}
```

---

### Example - With Hooks - useState

```typescript
import React, { useState } from "react";

const App = () => {
  const [name, handleName] = useState("Kawamura");
  const [favoriteFood, handleFoodName] = useState("curry");
  return (
    <>
      <input value={name} onChange={handleNameChange} />
      <input value={favoriteFood} onChange={handleFavoriteFood} />
    </>
  );
};
```

---

### Hooks - useState

---

> ### Example - useEffects

Reviewer: useEffect

---

> ### Example - useContexts

Reviewer: useContext

---

> ### Example - useReducers

Reviewer: useReducer

---

### For more information

> <!-- ![Dan_Hook_Presentation](./assets/youtube_hooks.png) -->

Reviewer: Maybe we can drop it?

Let's say we want to fetch some posts from a server with this API: https://jsonplaceholder.typicode.com/posts.

> Using the Context API, we can natively implement this kind of solution:

Reviewer: You don't mind, can you change it : "Using the Context API, a native implementation could be:"
